# <center>House price prediction with machine learning models 🤖</center>

![house](image.png)

## Contents

- [Getting the data](#getting-the-data)
- [Data analysis](#data-analysis)
- [Data cleaning](#data-cleaning)
- [Data visualization](#data-visualization)
- [Setting up the models](#setting-up-the-models)
- [Training the models](#training-the-models)

## Data description 📝

- `id` - Unique ID for every individual row entry  
- `date` - Date the house was sold  
- `price` - Price of the house  
- `bedrooms` - Number of bedrooms  
- `bathrooms` - Number of bathrooms  
- `sqft_living` - Square footage of the house  
- `sqft_lot` - Square footage of the lot  
- `floors` - Number of floors  
- `waterfront` - Whether the house has a view to a waterfront  
- `view` - How good the view of the property is  
- `condition` - Condition of the house  
- `grade` - Grade given to the house based on the overall construction and design  
- `sqft_above` - Square footage of house apart from the basement  
- `sqft_basement` - Square footage of the basement  
- `yr_built` - Year the house was built  
- `yr_renovated` - Year the house was renovated  
- `zipcode` - Zipcode of the house  
- `lat` - Latitude coordinate  
- `long` - Longitude coordinate  
- `sqft_living15` - Living room area in 2015(implies-- some renovations) This might or - might not have affected the lotsize area  
- `sqft_lot15` - LotSize area in 2015(implies-- some renovations)  

<hr>

In [35]:
import numpy as np
import pandas as pd
import seaborn as sn
import matplotlib.pyplot as plt

## Getting the data

In [36]:
# Create a dataframe from the csv file
df = pd.read_csv('kc_house_data.csv')

In [37]:
# Shape of the dataframe
df.shape

(21613, 21)

<hr>

## Data exploration

In [38]:
# Display the first 5 rows of the dataframe
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [39]:
# Display the last 5 rows of the dataframe
df.tail()

id             date     price  bedrooms  bathrooms  \
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[5 rows x 21 columns]

In [40]:
# Sample random 10 rows from the dataframe
df.sample(10)

id             date      price  bedrooms  bathrooms  \
21082  8562900430  20140718T000000   800000.0         4       2.50   
10426  6746700565  20141023T000000   447000.0         2       1.00   
9693   8644300200  20140605T000000   555000.0         4       2.75   
12279  2313900810  20150402T000000   610000.0         4       2.00   
17248   725069102  20150330T000000   650000.0         3       2.25   
16567  7603100095  20141110T000000  1260000.0         3       3.00   
2079   1336800160  20140605T000000   875000.0         5       2.50   
584    5419800510  20141117T000000   268500.0         4       1.75   
19849  7708210070  20140617T000000   535000.0         4       2.75   
12868  4014400237  20140523T000000   132500.0         3       1.00   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
21082         3691     11088     2.0           0     1  ...      8   
10426          850      2700     1.0           0     0  ...      6   
9693          2020     10720     1.0           0     0  ...      8   
12279         2220      5821     1.5           0     0  ...      7   
17248         2180     60112     2.0           0     0  ...      8   
16567         3230      8625     2.0           0     3  ...     10   
2079          2920      5568     2.0           0     0  ...      8   
584           1420      7500     1.0           0     0  ...      7   
19849         3070      7201     2.0           0     0  ...      9   
12868         1080     10500     1.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
21082        3691              0      2013             0    98074  47.6122   
10426         850              0      1924             0    98105  47.6684   
9693         1420            600      1976             0    98052  47.6373   
12279        1380            840      1916             0    98116  47.5723   
17248        2180              0      1976             0    98053  47.6723   
16567        2220           1010      1998             0    98116  47.5620   
2079         2320            600      1906             0    98112  47.6265   
584          1080            340      1981             0    98031  47.4025   
19849        3070              0      2006             0    98059  47.4897   
12868        1080              0      1967             0    98001  47.3200   

          long  sqft_living15  sqft_lot15  
21082 -122.059           3190       11270  
10426 -122.316           1630        3000  
9693  -122.104           2190       10164  
12279 -122.382           1850        5000  
17248 -122.082           2060      120225  
16567 -122.404           2330        6022  
2079  -122.312           2970        5568  
584   -122.176           1500        7260  
19849 -122.147           2880        8364  
12868 -122.278           1200        9607  

[10 rows x 21 columns]

In [41]:
# Display the columns of the dataframe
df.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [42]:
# Display info and the datatypes of the columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

### Data shows the following:

- There are 21,613 rows or entries in the dataset
- There are 21 columns or features in the dataset
- There are different data types in the dataset, including `int64`, `float64`, and `object`
- There are no missing values in the dataset
- There are 15,458 unique values in the `id` column
- Memory usage of the dataset is 3.5+ MB

## Data cleaning

In [43]:
# Search for missing values
df.isnull().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

## Data cleaning

In [ ]:
# Search for missing values
df.isnull().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

In [44]:
# Display the statistics of the dataframe. This method provides a summary of the numerical attributes like `count`, `mean`, `min`, `max` and `std`.
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

In [45]:
df.price.describe()

count    2.161300e+04
mean     5.400881e+05
std      3.671272e+05
min      7.500000e+04
25%      3.219500e+05
50%      4.500000e+05
75%      6.450000e+05
max      7.700000e+06
Name: price, dtype: float64

In [46]:
# Numeric Features
numeric_features = df.select_dtypes(['int', 'float']).columns
numeric_features

Index(['id', 'price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot',
       'floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long',
       'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [47]:
# Categorical Features
categorical_features = df.select_dtypes('object').columns
categorical_features

Index(['date'], dtype='object')

In [48]:
print(f'Number of `Numerical` features: {len(numeric_features)}')
print(f'Number of `Categorical` features: {len(categorical_features)}')
print(f'Total features: {len(numeric_features) + len(categorical_features)}')

Number of `Numerical` features: 20
Number of `Categorical` features: 1
Total features: 21


In [49]:

# Find unique values in dataframe
print(f"Total records in the dataframe: {len(df)}")
for col in numeric_features:
    print(f'Unique values in {col} are: {len(df[col].unique())}')

Total records in the dataframe: 21613
Unique values in id are: 21436
Unique values in price are: 4028
Unique values in bedrooms are: 13
Unique values in bathrooms are: 30
Unique values in sqft_living are: 1038
Unique values in sqft_lot are: 9782
Unique values in floors are: 6
Unique values in waterfront are: 2
Unique values in view are: 5
Unique values in condition are: 5
Unique values in grade are: 12
Unique values in sqft_above are: 946
Unique values in sqft_basement are: 306
Unique values in yr_built are: 116
Unique values in yr_renovated are: 70
Unique values in zipcode are: 70
Unique values in lat are: 5034
Unique values in long are: 752
Unique values in sqft_living15 are: 777
Unique values in sqft_lot15 are: 8689


In [50]:
# Remove columns that are not required
df.drop(['id', 'date'], axis=1, inplace=True)
df.head()

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0  221900.0         3       1.00         1180      5650     1.0           0   
1  538000.0         3       2.25         2570      7242     2.0           0   
2  180000.0         2       1.00          770     10000     1.0           0   
3  604000.0         4       3.00         1960      5000     1.0           0   
4  510000.0         3       2.00         1680      8080     1.0           0   

   view  condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  \
0     0          3      7        1180              0      1955             0   
1     0          3      7        2170            400      1951          1991   
2     0          3      6         770              0      1933             0   
3     0          5      7        1050            910      1965             0   
4     0          3      8        1680              0      1987             0   

   zipcode      lat     long  sqft_living15  sqft_lot15  
0    98178  47.5112 -122.257           1340        5650  
1    98125  47.7210 -122.319           1690        7639  
2    98028  47.7379 -122.233           2720        8062  
3    98136  47.5208 -122.393           1360        5000  
4    98074  47.6168 -122.045           1800        7503

## Data visualization

Contenido de la sección "Data visualization"

## Setting up the models

Contenido de la sección "Setting up the models"

## Training the models

Contenido de la sección "Training the models"